fedbiomed/transport/client.py

```diff
@@ -485,113 +575,120 @@ def __init__(
         self._retry_count = 0
         self._retry_msg = None
         self._stub_type = _StubType.NO_STUB
+        self._task_name = "sender"
+        self._want_retry_grpc_error = True
 
-    async def _listen(self, callback: Optional[Callable] = None) -> None:
-        """Listens for the messages that are going to be sent to researcher.
+
+    async def _handle_after_process(self,
+                                    status: ClientStatus,
+                                    want_retry: bool = False,
+                                    want_reconnect: bool = False,
+                                    want_detailed_log: bool = False,
+                                    post_noretry_function: Optional[Callable] = None,
+                                    *args
+                                    ):
+        """Actions after each tentative to send a message, successful or not
 
         Args:
-            callback: Callback to execute once a task is received
+            status: new gRPC client status to set
+            want_retry: if True (and MAX_SEND_RETRIES is not exceeded) then re-send message
+            want_reconnect: unused
+            want_detailed_log: if True (or MAX_SEND_RETRIES is exceeded) then log details about this message
+            post_noretry_function: optional final function to execute if message is not re-sent.
+                If None, no final function is executed
+            args: arguments for `post_noretry_function`
+        """
+        await self._on_status_change(status)
+
+        if want_detailed_log or self._retry_count >= MAX_SEND_RETRIES:
+            # Extract useful information for detailed log without assumption on message structures
+            # to cover possible bug cases
+            if isinstance(self._retry_msg, dict) and {'stub', 'message'} <= self._retry_msg.keys():
```

Comment:
Hi @mvesin,
I see your point. However, this can be implemented another way because this not the only place that we access `stub` and `message`. If you want to make sure that this part should be modified  if any change is done in the item structure  I would prefer to define dataclass that has `stub` and `message` attribute. This dataclass can be instantiated where the item is put in the queue. So it will ensure that every item that is put in the queue will respect the format.  Thanks to this implementation we will be sure that the rest of the execution will respect the format. Plus, you can add specific unit tests that builds this dataclass and pass it to the queue. This will ensure that the tests will fail if someone changes the datatype.

Reply:
Thank you @srcansiz for your answer.

However, I think adding a dataclass adds un-needed complexity and does not really address the point I mentioned.

As we cannot really converge views on that (minor) point, I will then push your initial requirement (remove test for existence of `stub` and `message` elements) to reach conclusion on this point.